core/src/utils/mod.rs

```diff
@@ -131,20 +480,126 @@ fn debug_assert_is_checked() {
 
 /// Utility to convert a sequence of bytes to hex.
 pub fn to_hex(bytes: &[u8]) -> Result<String, fmt::Error> {
+    use core::fmt::Write;
+
     let mut s = String::with_capacity(bytes.len() * 2);
+    write!(s, "{:x}", DisplayHex(bytes))?;
+    Ok(s)
+}
 
-    for byte in bytes {
-        write!(s, "{byte:02x}")?;
+/// A display helper for formatting a slice of bytes as hex
+/// with different options using Rust's builtin format language
+pub struct DisplayHex<'a>(pub &'a [u8]);
```

Comment:
Do we need this struct primarily so that we can implement `PrettyPrint` on it?

Separate from this PR, we actually have hex manipulation code in a few places (e.g., in [miden-crypto](https://github.com/0xPolygonMiden/crypto/blob/main/src/utils/mod.rs#L32)). We should think about how to consolidate it in one place. Maybe it makes sense to create a `miden-utils` crate which would have things like hex formatting, pretty printing, and a few other things in it.

Reply:
I'd maybe go with `miden-formatting` or something that is _just_ code for handling pretty-printing/formatting of text/output. That will make it more broadly useful, but also fairly narrowly scoped so it doesn't turn into a big bucket of random things.

But to answer your question, it's not just for implementing `PrettyPrint`, it's also so that you can use all of the `core::fmt` options to configure the output, e.g. `{:#x}` vs `{:x}` to control whether there is a leading `0x` prefix. This is preferable to things like `write_hex_bytes` which don't provide this configurability, and are less convenient to use in things like an implementation of `PrettyPrint`, where we aren't actually rendering to a `core::fmt::Formatter` yet, but we still need the display representation for some object that renders as hexadecimal bytes.